packages/malloy-vscode/src/common/connection_manager.ts

```diff
@@ -11,38 +11,170 @@
+  LookupConnection,
   TestableConnection,
 } from "@malloydata/malloy";
 import {
   ConnectionBackend,
   ConnectionConfig,
-  getDefaultIndex,
 } from "./connection_manager_types";
 import { getPassword } from "keytar";
 
+const DEFAULT_CONFIG = Symbol("default-config");
+let ConnectionPool: Record<string, TestableConnection> = {}
+
+interface ConfigOptions {
+  workingDirectory: string;
+  rowLimit?: number;
+  usePool?: boolean;
+}
+
+const getConnectionForConfig = async (
+  connectionConfig: ConnectionConfig,
+  { workingDirectory, rowLimit, usePool }: ConfigOptions = { workingDirectory: "/" }
+): Promise<TestableConnection> => {
+  let connection: TestableConnection;
+  if (usePool && ConnectionPool[connectionConfig.name]) {
+    return ConnectionPool[connectionConfig.name];
+  }
+  switch (connectionConfig.backend) {
+    case ConnectionBackend.BigQuery:
+      connection = new BigQueryConnection(
+        connectionConfig.name,
+        () => ({ rowLimit }),
+        {
+          defaultProject: connectionConfig.projectName,
+          serviceAccountKeyPath: connectionConfig.serviceAccountKeyPath,
+          location: connectionConfig.location,
+        }
+      );
+      if (usePool) {
+        ConnectionPool[connectionConfig.name] = connection;
+      }
+      break;
+    case ConnectionBackend.Postgres: {
+      const configReader = async () => {
+        let password;
+        if (connectionConfig.password !== undefined) {
+          password = connectionConfig.password;
+        } else if (connectionConfig.useKeychainPassword) {
+          password =
+            (await getPassword(
+              "com.malloy-lang.vscode-extension",
+              `connections.${connectionConfig.id}.password`
+            )) || undefined;
+        }
+        return {
+          username: connectionConfig.username,
+          host: connectionConfig.host,
+          password,
+          port: connectionConfig.port,
+          databaseName: connectionConfig.databaseName,
+        };
+      };
+      connection = new PostgresConnection(
+        connectionConfig.name,
+        () => ({ rowLimit }),
+        configReader
+      );
+      if (usePool) {
+        ConnectionPool[connectionConfig.name] = connection;
+      }
+      break;
+    }
+    case ConnectionBackend.DuckDB: {
+      if (!isDuckDBAvailable) {
+        throw new Error("DuckDB is not available.");
+      }
+      try {
+        connection = new DuckDBConnection(
+          connectionConfig.name,
+          ":memory:",
+          connectionConfig.workingDirectory || workingDirectory
+        );
+      } catch (error) {
+        console.log("Could not create DuckDB connection:", error);
+        throw error;
+      }
+      break;
+    }
+  }
+
+  // Retain async signature for future reference
+  return Promise.resolve(connection);
+};
+
+export class DynamicConnectionLookup implements LookupConnection<Connection> {
+  connections: Record<string | symbol, Promise<Connection>> = {};
```

Comment:
Could even do `string | typeof DEFAULT_CONFIG` here if you wanted to get really fancy.